main_window/read_thread.py

```diff
@@ -59,23 +60,26 @@ def run(self):
                     else:
                         continue
 
-            # Convert from immutable bytearray to List[int]
-            byte_list = list(data)  # list of ints ranging from 0 - 255. Equivalent to doing [x for x in data]
+            byte_stream: BytesIO = BytesIO(data)
+            curr_offset = 0
 
-            # loop that quickly runs through entire data list and extracts subpackets where possible
-            while len(byte_list) > 0:
-                parsed_data: Dict[int, any] = {}  # generally any is floats and ints
-                length: int = 0
+            # Get length of bytes (without using len(data) for decoupling)
+            byte_stream.seek(0, SEEK_END)
+            end = byte_stream.tell()
+            byte_stream.seek(0)
+
+            # Iterate over stream to extract subpackets where possible
+            while byte_stream.tell() < end:
                 try:
-                    parsed_data, length = self.radioController.extract(byte_list)
-                    self.rocketData.addBundle(parsed_data)
+                    parsed_data: Dict[int, any] = self.packet_parser.extract(byte_stream)
+                    self.rocket_data.addBundle(parsed_data)
+                    curr_offset = byte_stream.tell()
 
                     # notify UI that new data is available to be displayed
                     self.sig_received.emit()
                 except Exception as e:
                     LOGGER.exception("Error decoding new data!")  # Automatically grabs and prints exception info
-
-                del byte_list[0:length]
+                    byte_stream.seek(curr_offset+1)
```

Comment:
So the reasoning is that the extractor and consumer of the byte_stream will have moved the offset somewhere unknown, and as caller we should not depend on it being anywhere in particular.

As for dropping entirely, I'm not sure about what our error, correctness and dropping policy is, maybe we should discuss further

Reply:
> So the reasoning is that the extractor and consumer of the byte_stream will have moved the offset somewhere unknown, and as caller we should not depend on it being anywhere in particular.

Because we're using the stream, isnt curr_offset going to be in a bad place if we throw mid-parse? So curr_offset+1 will likely be just as bad.

IIRC XBee's spec discard a packet if there's something wrong in it. Since each iteration of the main thread loop `data` is effectively a packet (superpacket? packetgroup? whatever we want to call it, maybe we should rename it for clarity), we can take a page out of XBee's book and just discard the the whole thing if there's something wrong with it.

Reply:
Sure. That +1 strategy was an artifact of the old implementation I guess, so makes sense to follow XBee logic.

Reply:
You're right though. It would be nice to treat it as an infinite stream and not depend on the fact that we're getting chunks. Just not sure how atm.